Code/Framework/AzCore/AzCore/DOM/DomValue.h

```diff
@@ -0,0 +1,385 @@
+        Object::ConstIterator FindMember(KeyType name) const;
+        Object::ConstIterator FindMember(AZStd::string_view name) const;
+
+        Value& MemberReserve(size_t newCapacity);
+        bool HasMember(KeyType name) const;
+        bool HasMember(AZStd::string_view name) const;
+
+        Value& AddMember(KeyType name, const Value& value);
+        Value& AddMember(AZStd::string_view name, const Value& value);
+        Value& AddMember(KeyType name, Value&& value);
+        Value& AddMember(AZStd::string_view name, Value&& value);
+
+        void RemoveAllMembers();
+        void RemoveMember(KeyType name);
+        void RemoveMember(AZStd::string_view name);
+        Object::Iterator RemoveMember(Object::Iterator pos);
+        Object::Iterator EraseMember(Object::ConstIterator pos);
+        Object::Iterator EraseMember(Object::ConstIterator first, Object::ConstIterator last);
+        Object::Iterator EraseMember(KeyType name);
+        Object::Iterator EraseMember(AZStd::string_view name);
+
+        Object::ContainerType& GetMutableObject();
+        const Object::ContainerType& GetObject() const;
+
+        // Array API (also used by Node)...
+        Value& SetArray();
+
+        size_t Size() const;
+        size_t Capacity() const;
+        bool Empty() const;
+        void Clear();
+
+        Value& operator[](size_t index);
+        const Value& operator[](size_t index) const;
+
+        Value& MutableAt(size_t index);
+        const Value& At(size_t index) const;
+
+        Array::ConstIterator Begin() const;
+        Array::ConstIterator End() const;
+        Array::Iterator Begin();
+        Array::Iterator End();
+
+        Value& Reserve(size_t newCapacity);
+        Value& PushBack(Value value);
+        Value& PopBack();
+
+        Array::Iterator Erase(Array::ConstIterator pos);
+        Array::Iterator Erase(Array::ConstIterator first, Array::ConstIterator last);
+
+        Array::ContainerType& GetMutableArray();
+        const Array::ContainerType& GetArray() const;
+
+        // Node API (supports both object + array API, plus a dedicated NodeName)...
+        void SetNode(AZ::Name name);
+        void SetNode(AZStd::string_view name);
+
+        AZ::Name GetNodeName() const;
+        void SetNodeName(AZ::Name name);
+        void SetNodeName(AZStd::string_view name);
+        //! Convenience method, sets the first non-node element of a Node.
+        void SetNodeValue(Value value);
+        //! Convenience method, gets the first non-node element of a Node.
+        Value GetNodeValue() const;
+
+        Node& GetMutableNode();
+        const Node& GetNode() const;
+
+        // int API...
+        int64_t GetInt64() const;
+        void SetInt64(int64_t);
+        int32_t GetInt32() const;
+        void SetInt32(int32_t);
+
+        // uint API...
+        uint64_t GetUint64() const;
+        void SetUint64(uint64_t);
+        uint32_t GetUint32() const;
+        void SetUint32(uint32_t);
```

Comment:
Providing 32-bit API function is cluttering the API.
The user can just easily perform the numeric_cast of the 64-bit API if they desire smaller values from the `GetUint64` and `GetInt64` methods?
The 32-bit Set methods are just fluff

Furthermore this API is incomplete, if was going to provide 32-bit functions.
Either it should have provided all the integer size types are part of the API(8-bit, 16-bit, 32-bit or 64-bit) or provide the only largest type(64-bit). Go all the way or none of the way, but not in-between.

Reply:
Even in just writing unit tests I've found the ergonomics of having a 32 bit API pretty helpful versus constantly casting, so I'm inclined to leave this as-is. I've gone ahead and added 16 and 8 bit APIs for completeness to address that point - if you still feel strongly about ripping these out, hit me up and I'll likely capitulate 😄

Reply:
I am OK as long as the 8-bit and 16-bit APIs were also added.

Reply:
To a degree this is more of a legacy thing. Several libraries for these formats, like RapidJSON, were created when 32-bit CPUs were still in use so they started with 32 bit integers and later added 64-bit integers. Having 32 and 64 bit versions would make it easier to work with these libraries. That said, it's not strictly necessary, so if the choice is between using 64-bit only or adding 8 and 16 bit calls I would keep things simple and just simplify to 64-bit values only.

Reply:
I'm finding when writing tests it's been pretty convenient to have the 32 bit API on Value. The visitor API is still strictly 64-bit, so it shouldn't encroach on the rest of the API (e.g. the serializers and parsers doesn't need to go to/from N-bit integral types) - if we feel really strongly about this it can be removed, but I expect it to cause some ergonomic issues when trying to do things like:
```
    int x;
    Dom::Value v(x);
```